### ISSUE TITLE
(ApisMessageVo m = (ApisMessageVo)messageVo;)Downcast is not working at Runtime

### ISSUE THREAD

**Issue by RamakriJaya** [US-CBP/GTAS#71]
We have refactored locally the gtas-loader , gtas-commons and gtas-parsers to load customer related data .
Also we have modified the MessagegeneratorApp to create data that is relevant for the customer table.
The data is generated. We have considered mostly only Mandatory segments in each group for creating data.
When we are adding the TDT and LOC from segment 2 and segment 3 and omitt COM which is conditional in group1, the parser passes but the customers list is empty.On debugging , we found that in 
ApisMessageService.java , inside the load method downcast is performed.but customers value is empty when we do a m.getCustomers here.
Hence the data is not loaded in the customers table.
But in the message table has a record created 

ApisMessageVo m = (ApisMessageVo)messageVo; //Downcast is not working for customers

Code snippets are below

/----------------------------------------------------------------------/

@Service
public class ApisMessageService extends MessageLoaderService {
    private static final Logger logger = LoggerFactory.getLogger(ApisMessageService.class);

```
@Autowired
private ApisMessageRepository msgDao;

private ApisMessage apisMessage;

@Override
public List<String> preprocess(String message) {
    logger.info("Print the message array:" + Arrays.asList(message));
    return Arrays.asList(message);
}

@Override
public MessageVo parse(String message) {
    logger.info("Parse string is :" + message);
    apisMessage = new ApisMessage();
    apisMessage.setCreateDate(new Date());
    apisMessage.setStatus(MessageStatus.RECEIVED);
    apisMessage.setFilePath(filePath);

    MessageVo vo = null;
    try {            
        EdifactParser<ApisMessageVo> parser = null;
        if (isUSEdifactFile(message)) {
            parser = new PaxlstParserUSedifact();
        } else {
            parser = new PaxlstParserUNedifact();                
        }

        vo = parser.parse(message);
        logger.info("After parsing the value in VO set  :" + vo.getRaw());
        loaderRepo.checkHashCode(vo.getHashCode());
        apisMessage.setRaw(LobUtils.createClob(vo.getRaw()));
        logger.info("The value that is set in apisMessage customers object :" + apisMessage.getCustomers());

        apisMessage.setStatus(MessageStatus.PARSED);
        apisMessage.setHashCode(vo.getHashCode());
        EdifactMessage em = new EdifactMessage();
        em.setTransmissionDate(vo.getTransmissionDate());
        em.setTransmissionSource(vo.getTransmissionSource());
        em.setMessageType(vo.getMessageType());
        em.setVersion(vo.getVersion());
        apisMessage.setEdifactMessage(em);
    } catch (Exception e) {
        handleException(e, MessageStatus.FAILED_PARSING);
        return null;
    } finally {
        logger.info("inside the finally:" + apisMessage.getStatus() );
        createMessage(apisMessage);
    }
    logger.info("parse status:" + apisMessage.getStatus() );
    return vo;
}

@Override
public boolean load(MessageVo messageVo) {
    boolean success = true;
    logger.info("The initial values inside apisMessage is : " +apisMessage.getCustomers() );//Check the initial value.
    logger.info("The initial values inside apisMessage is : " +apisMessage.getRaw() );
    logger.info("The value that is passed in messageVo :" + messageVo.getRaw());

    try {
        ApisMessageVo m = (ApisMessageVo)messageVo;
        //Replaced with code below check if it works
        //TODO check if explicit setting works as downcasting is not having the appropriate values
```

//            ApisMessageVo m = new ApisMessageVo();
//            MessageVo mVo = m;
//            ApisMessageVo m2 = (ApisMessageVo)mVo;
            logger.info("After casting messageVo to apisMessageVo--VO :" + m.getCustomers());

//            loaderRepo.processReportingParties(apisMessage, m.getReportingParties());
            loaderRepo.processCustomers(m.getCustomers() ,apisMessage.getCustomers());
            apisMessage.setStatus(MessageStatus.LOADED);

```
    } catch (Exception e) {
        success = false;
        handleException(e, MessageStatus.FAILED_LOADING);
    } finally {
        success &= createMessage(apisMessage);     

    }
    logger.info("load status:" + apisMessage.getStatus() );
    return success;
}  

private void handleException(Exception e, MessageStatus status) {
```

//        apisMessage.setFlights(null);
        apisMessage.setStatus(status);
        String stacktrace = ErrorUtils.getStacktrace(e);
        apisMessage.setError(stacktrace);
        logger.error(stacktrace);
    }

```
@Transactional
private boolean createMessage(ApisMessage m) {
    logger.info("inside the createMessage:" + m.getCustomers() );
    boolean ret = true;
    try {
        msgDao.save(m);
        indexer.indexApis(m);
    } catch (Exception e) {
        ret = false;
        handleException(e, MessageStatus.FAILED_LOADING);
        msgDao.save(m);
    }
    logger.info("after executing the createMessage:" + ret );
    return ret;
}

private boolean isUSEdifactFile(String msg) {
    return (msg.contains("CDT") || msg.contains("PDT"));
}
```

}

/-------------------------------------------------------------------------------------/

```
private List<CustomerDto> buildCustomerList(List<CustomerDto> custList){

    System.out.println("################"+custList.size()); 
        int j=GenUtil.getRandomNumber(99)+100;
        System.out.println("################"+j); 
        for(int i=1;i<=j;i++){
            CustomerDto pax= new CustomerDto();
            pax.setId(i);
            if(i == 2 ){//watch list data
                pax=GenUtil.getCustomersDto();
            }
            else{
                pax.setFirstName(GenUtil.getFirstName());
                pax.setLastName(GenUtil.getLastName());
                pax.setDob(GenUtil.getBirthDate());
                pax.setDocNumber(GenUtil.getDocumentNumber());
                System.out.println("First Name"+pax.getFirstName());
                System.out.println("Last Name"+pax.getLastName());

            }

            custList.add(pax);
        }

        System.out.println("################"+custList.size()); 
        return custList;

}
```

Attaching the data file generated using messageGeneratorApp.
[apis7751.txt](https://github.com/US-CBP/GTAS/files/484216/apis7751.txt)

In Summary,
1>The change we have made is using the Customer name instead of Passenger for value objects. All names have been changed appropriately to ensure there are no build errors.
2>When reading from the same data from the edi file, this is causing the downcasting to not take effect and parser to not function. 
3>Consequently the loader also fails when loading this data to the DB. 

Please let us know if we are missing anything here.

Thanks
Jayabalambika.R

**Comment by copenhafer** [comment]
Jayabalambika -- I'm a little confused.  Are you saying there's a problem with the original code as is?  Or only a problem with the changes you've made w/ the addition of the Customer object?
